allennlp/nn/beam_search.py

```diff
@@ -25,6 +27,389 @@
+        sampled_indices = torch.multinomial(
+            normalized_top_k_probs, per_node_beam_size, replacement=self.with_replacement
+        )
+        # Convert `sampled_indices` back to indices in the original `log_probs` tensor.
+        # shape: (batch_size, per_node_beam_size)
+        indices = top_k_indices.gather(-1, sampled_indices)
+        return log_probs.gather(1, indices), indices, state
+
+
+@Sampler.register("top-p")
+class TopPSampler(Sampler):
+    """
+    A `Sampler` which redistributes the probability mass function for nodes among
+    the top choices with a cumulative probability of at least `p`, then samples from that subset
+    after re-normalizing the probabilities.
+
+    Beams are sampled in the default, deterministic way.
+
+    # Parameters
+
+    p : `float`, optional (default = `0.9`)
+        The cumulative probability cutoff threshold. A higher value of `p` will result in more possible
+        examples to sample from.
+    temperature : `float`, optional (default = `1.0`)
+        A `temperature` below 1.0 produces a sharper probability distribution and a `temperature`
+        above 1.0 produces a flatter probability distribution.
+    with_replacement : `bool`, optional, (default = `False`)
+        If set to `True`, samples will be selected with replacement from the top choices.
+    """
+
+    def __init__(
+        self,
+        p: float = 0.9,
+        temperature: float = 1.0,
+        with_replacement: bool = False,
+    ):
+        if p < 0.0 or p > 1.0:
+            raise ValueError("p must be a non-negative float no greater than 1.0")
+        self.p = p
+        self.temperature = temperature or 1.0
+        self.with_replacement = with_replacement
+
+    @overrides
+    def sample_nodes(
+        self, log_probs: torch.Tensor, per_node_beam_size: int, state: StateType
+    ) -> Tuple[torch.Tensor, torch.Tensor, StateType]:
+        # Calculate filter value
+        filter_val = min_value_of_dtype(log_probs.dtype)
+
+        # First apply temperature coefficient:
+        if self.temperature != 1.0:
+            _log_probs = torch.nn.functional.log_softmax(log_probs / self.temperature, dim=-1)
+        else:
+            _log_probs = log_probs
+
+        # Sort the probabilities in descending order to then find cumulative sum
+        log_probs_descending, sorting_indices = torch.sort(_log_probs, descending=True)
+
+        # shape: (batch_size, num_classes)
+        probabilities_descending = log_probs_descending.exp()
+        probabilities_summed = torch.cumsum(probabilities_descending, dim=-1)
+
+        # Create a mask for filtering out probabilities that don't make the top `p`.
+        # shape: (batch_size, num_classes)
+        exclusion_mask = probabilities_summed >= self.p
+
+        # We want to include the firt index where probabilities_summes >= p, so we shift over one.
+        exclusion_mask[..., 1:] = exclusion_mask[..., :-1].clone()
+        exclusion_mask[..., 0] = False
```

Comment:
How about this:

```suggestion
        # Create a mask for filtering out probabilities that don't make the top `p`.
        # shape: (batch_size, num_classes)
        exclusion_mask = probabilities_summed > self.p
        # Make sure there's at least `per_node_beam_size` options.
        exclusion_mask[..., :per_node_beam_size] = False
```

Reply:
I think we actually need both. If we don't shift the mask then then the cumulative sum is less than p, i.e.
probs = [0.5, 0.31, 0.19] , p = 0.8
cumsum = [0.5, 0.81, 1.0]
mask: [False, True, True]
softmax: [1.0, 0.0, 0.0]

Reply:
Good point. How about this then:

```suggestion
        # Create a mask for filtering out probabilities that don't make the top `p`.
        # shape: (batch_size, num_classes)
        exclusion_mask = probabilities_summed >= self.p

        # We want to include the first index where probabilities_summed >= p, so we shift over one.
        exclusion_mask[..., 1:] = exclusion_mask[..., :-1].clone()

        # We also want to make sure we have at least `per_node_beam_size` options.
        exclusion_mask[..., :per_node_beam_size] = False
```

Reply:
Do we want to allow the client to sample  with replacement from < per_node_beam_size options?

Reply:
i.e. maybe they sample the 10 examples from the top 8 cumulative probabilities

Reply:
Possibly. So maybe only fill up to `per_node_beam_size` as `False` if `with_replacement` is `False`, otherwise only fill the first as `False`.

Reply:
If you don't think we should build that functionality in though I can clean it up!

Reply:
I think it's okay to have that.

Reply:
oops sorry I forgot to refresh and missed your comment! And yes that's the change I ended up making, so it should be all good